### ISSUE TITLE
AuthenticateSelf

### ISSUE THREAD

**Issue by LostSoulfly** [hexafluoride/Bifrost#7]
I see that you added the option AuthenticateSelf to, from what I can tell, allow each side to authenticate the connection without the need for passing a public key (.ca) out of band. I'm glad this option exists, but I don't see an easy way to store the other end's public key for future use.

How difficult is this to add in? I'd like to do an initial connection to exchange keys and then store the keys to be used again and ensure the connection is secure and there is no MITM. In this way, only the initial connection has the potential for compromise. I plan to verify the remote side's public key through other means to verify it's safe, but I'd like to exchange the keys automatically at least once.

Thanks!

**Comment by LostSoulfly** [comment]
I believe I've adequately implemented it to suite my needs here: https://github.com/LostSoulfly/BifrostLSF

I'm not a very experienced programmer so I won't be create PRs, but if you see anything I might have done wrong do let me know!

**Comment by hexafluoride** [comment]
Hello, I'm unfortunately seeing this issue only now (blame gmail for being dumb about notifications and blame me for not checking my GitHub). I'll check out your changes when I wake up in a few hours and merge them back here if that sits okay with you? Sorry for being this late, and thanks for your interest in my library!

**Comment by LostSoulfly** [comment]
Hey, no worries. I've not really implemented threading in the past so it was a great learning experience for me. I'm totally okay with you looking at my changes and cherry picking anything that you think might be worth using. I made several changes/additions but only really to the things that I plan to use -- TCP and the Server/client/Encrypted links.

If you decide to implement these sorts of features into Bifrost more elegantly I'd be happy to make the switch. The biggest things for me are multi-user server, a client that runs on its own thread, allowing unknown certificates but remembering them for future authentication, and a transparent way to serialize/deserialize a custom message/class and rebuild it on the other end as opposed to ending/receiving a byte array.

I appreciate the effort you put into the library and plan to use some version of it in several projects in the future!

For reference - I made changes to my Bifrost fork here: https://github.com/LostSoulfly/BifrostLSF
And wrote a separate library to use my Bifrost fork here: https://github.com/LostSoulfly/BifrostExtended
